README.md
# Virtualmin API
![GitHub release (latest by date)](https://img.shields.io/github/v/release/fintech-systems/virtualmin-api) [![Build Status](https://app.travis-ci.com/fintech-systems/virtualmin-api.svg?branch=main)](https://app.travis-ci.com/fintech-systems/virtualmin-api) ![GitHub](https://img.shields.io/github/license/fintech-systems/virtualmin-api)

A Virtualmin API designed to run standalone or as part of a Laravel Application

Requirements:

- PHP 8.0
- A running Virtualmin server, using the [Remote API](https://www.virtualmin.com/documentation/developer/http/)

Features:

- List Domains (Virtualmin output)
- Get Domains (User-Friendly Output)
- Change Plan
- List Plans (Virtualmin output)
- Get Plans (User-Friendly Output)
- Create Domain

## Get Domains

Framework Agnostic PHP:

```php
use \FintechSystems\VirtualminApi\VirtualminApi;
$api = new VirtualminApi(
        'hostname' => 'virtualmin.test',
        'username' => 'root',
        'password' => '********'
    );
$api->getDomains();
```

Laravel App:

```php
VirtualminApi::getDomains();
```

Output:

Array of all domains on a Virtualmin server.

## Change Plan

Framework Agnostic PHP:

```php
use \FintechSystems\VirtualminApi\VirtualminApi;
$api = new VirtualminApi;
$api->changePlan('example.com', 'New Plan Name');
```

Laravel App:

```php
VirtualminApi::changePlan('example.com', 'New Plan Name');
```

## Create Domain

Example:

```php
$api->createDomain([
    // New virtual server details
    'domain'                 => 'demo.example.com',
    'desc'                   => 'My demo website',
    'pass'                   => '***************',
    'template'               => 'WikiSuite 1',
    'plan'                   => 'WikiSuite 1',
    
    // Advanced options
    'email'                  => 'contact@example.com',
    'db'                     => 'demo_example_com',
    
    // Enabled features
    'features-from-plan'     => '',
    'virtualmin-tikimanager' => ''
]))
```

## Testing

We have tests! Use the command below to run the tests.

Live API calls will be made otherwise causing your tests to fail.

`vendor/bin/phpunit --exclude-group=live`

### Coverage reports

To regenerate coverage reports:

`XDEBUG_MODE=coverage ./vendor/bin/phpunit --coverage-html=tests/coverage-report`

See also `.travis.yml`

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Version Control

This application uses Semantic Versioning

https://semver.org/

## Local Editing

For local editing, add this to `composer.json`:

```json
"repositories" : [
        {
            "type": "path",
            "url": "../virtualmin-api"
        }
    ]
```
## Credits

- [Eugene van der Merwe](https://github.com/eugenevdm)
- [Fabio Montefuscolo](https://github.com/fabiomontefuscolo)
- [Marc Laporte](https://github.com/marclaporte)
- [Domeshow](https://github.com/Domeshow)

## Inspiration

This standalone package was inspired by video course by Marcel Pociot of BeyondCode:<br>
[PHP Package Development](https://beyondco.de/video-courses/php-package-development)

Before doing the video course I had developed many versions of the same thing but it was never standalone and as a consequence over the years it was difficult to maintain and quickly use in new projects.

## How to contribute

Please add issues and merge requests via GitHub. 

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Virtualmin API Change Log

All notable changes to `virtualmin-api` will be documented in this file.

## v0.1.5 - 2022-02-14

- FIX: Check if actual output is returned on getDomains() before looping (PR-23)

## v0.1.4 - 2022-02-06

- Add the domain ID to the domain info (PR #19)
- Update contributors in README

## v0.1.3 - 2022-02-03

- Change scope of runProgram to public

## v0.1.2 - 2022-01-26

- Added ability to create domain

## v0.1.1 - 2021-10-29

- Changed the WGET command that was run using a shell to native PHP CURL
- Improved README examples by adding name spacing and adding get and list plans as features
- Added Fabio as contributor

## 0.1.0 - 2021-09-10

First test version that include live server testing, offline testing, and list domains and get domains commands.

There is also a Laravel Service Provider

tests/Setup.php
<?php

namespace FintechSystems\VirtualminApi\Tests;

use Exception;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Dotenv\Dotenv;

class Setup extends TestCase
{
    // public $host;

    // public function __construct() {
    //     $this->host = $this->getHostInformation();
    // }

    private function dotEnvExists()
    {
        $dotenv = new Dotenv();

        try {
            $dotenv->load(__DIR__.'/../.env');
        } catch (Exception $e) {
            return false;
        }

        return true;
    }

    protected function getServerInformation()
    {
        if (! $this->dotEnvExists()) {
            $this->assertTrue(true);

            return false;
        }

        $server = [
            'hostname' => env('VIRTUALMIN_HOSTNAME'),
            'username' => env('VIRTUALMIN_USERNAME'),
            'password' => env('VIRTUALMIN_PASSWORD'),
        ];

        if (! $server['hostname']) {
            return false;
        }

        return $server;
    }
}


tests/ApiTest.php
<?php

namespace FintechSystems\VirtualminApi\Tests;

use FintechSystems\VirtualminApi\VirtualminApi;
use Symfony\Component\Dotenv\Dotenv;

class ApiTest extends Setup
{
    /** @test */
    public function it_can_read_an_env_testing_file_and_assign_it_to_an_array()
    {
        $dotenv = new Dotenv();
        $dotenv->load(__DIR__.'/../.env.testing');

        $hostname = [
            'hostname' => $_ENV['VIRTUALMIN_HOSTNAME'],
            'password' => $_ENV['VIRTUALMIN_PASSWORD'],
        ];

        $this->assertEquals('hostname.example.com', $hostname['hostname']);
    }

    /**
     * @test
     * @group live
     */
    public function it_retrieves_a_list_of_domains_on_a_live_server_and_shows_debugging_information()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'debug');

        $result = json_decode($api->listDomains());

        $this->assertEquals('success', $result->status);
    }

    /**
     * @test
     * @group live
     */
    public function it_retrieves_a_list_of_domains_on_a_live_server_and_saves_it_as_a_file()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'write_cache');

        $result = json_decode($api->listDomains());

        $this->assertEquals('success', $result->status);
    }

    /** @test */
    public function it_can_retrieve_a_list_of_domains_from_a_list_domains_json_example_on_disk()
    {
        $file = file_get_contents('storage/list-domains.json');

        $result = json_decode($file);

        $this->assertEquals('success', $result->status);
    }

    /** @test */
    public function it_retrieves_a_cached_response_for_list_domains_from_the_virtualmin_api()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'read_cache');

        $result = json_decode($api->listDomains());

        $this->assertEquals('success', $result->status);
    }

    /** @test */
    public function it_flattens_a_virtualmin_api_list_domains_command_and_returns_domains_ips_usernames_plans_and_statuses()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'read_cache');

        $result = $api->getDomains();

        $this->assertCount(6, $result);
    }

    /** @test */
    public function it_returns_all_plans_from_a_server_and_writes_them_to_disk()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'write_cache');

        $result = json_decode($api->listPlans());

        $this->assertCount(2, $result->data);
    }

    /** @test */
    public function it_returns_virtualmin_plans_in_an_user_friendly_format()
    {
        if (! $server = $this->getServerInformation()) {
            $this->assertTrue(true);

            return;
        }

        $api = new VirtualminApi($server, 'read_cache');

        $result = $api->getPlans();

        // dd($result);

        $this->assertCount(2, $result);
    }
}


tests/JokeFactoryTest.php
<?php

namespace Mpociot\ChuckNorrisJokes\Tests;

use Mpociot\ChuckNorrisJokes\JokeFactory;
use PHPUnit\Framework\TestCase;

class JokeFactoryTest extends TestCase
{
    /** @test */
    public function it_returns_a_random_joke()
    {
        $jokes = new JokeFactory([
            'This is a joke',
        ]);

        $joke = $jokes->getRandomJoke();

        $this->assertSame('This is a joke', $joke);
    }

    /** @test */
    public function it_calls_helo()
    {
        $jokes = new JokeFactory([
            'This is a joke',
        ]);

        $joke = $jokes->hello();

        $this->assertTrue(true);
    }
}


src/Api.php
<?php

namespace FintechSystems\VirtualminApi;

class Api
{
    private $host;
    private $debug;

    private $quiet = '--quiet';

    public function __construct($host = null, $debug = false)
    {
        $this->debug = $debug;

        if ($this->debug == true) {
            $this->quiet = '';
        }

        if ($host != null) {
            $this->host = $host;
        }
    }

    public function getDomains()
    {
        return $this->listDomains();
    }

    private function listDomains()
    {
        $program = 'list-domains';

        return $this->runProgram($program);
    }

    private function runProgram($program)
    {
        $host = $this->host['host'];
        $username = $this->host['username'] ?? 'root';
        $password = $this->host['password'];
        $port = $this->host['port'] ?? '10000';

        $command = "wget -O - $this->quiet --http-user='$username' --http-passwd='$password' --no-check-certificate 'https://$host:$port/virtual-server/remote.cgi?json=1&multiline&program=$program'";

        $result = shell_exec(
            $command
        );

        return json_decode($result);
    }

    // Deprecated
    public function test()
    {
        return 'The API was invoked.';
    }
}


src/VirtualminApi.php
<?php

namespace FintechSystems\VirtualminApi;

/**
 * An API that provides calls to Virtualmin.
 */
class VirtualminApi
{
    private $server;

    /**
     * Mode can be debug, write_cache or read_cache.
     */
    private $mode;

    /**
     * Wget won't produce output if this flag is set so for debugging, we clear it.
     */
    private string $quiet = '--quiet';

    private bool $debug;

    public function __construct($server = null, $mode = '')
    {
        $this->mode = $mode;

        if ($this->mode == 'debug') {
            $this->debug = true;
            $this->quiet = '';
        }

        if ($server != null) {
            $this->server = $server;
        }
    }

    /**
     * Format the raw output from the Virtualmin list-domains program into a more user-friendly format.
     */
    public function getDomains(): array
    {
        $domains = [];

        $output = json_decode($this->listDomains());

        if ($output) {
            foreach ($output->data as $domain) {
                $domains[] = [
                    'server'           => $this->server['hostname'],
                    'name'             => $domain->name,
                    'type'             => $this->getHostingType($domain),
                    'username'         => $domain->values->username[0] ?? '',
                    'password'         => $domain->values->password[0] ?? '',
                    'plan'             => $domain->values->plan[0],
                    'disk_space_used'  => $domain->values->server_byte_quota_used[0] ?? 0,
                    'disk_space_limit' => $this->getDiskSpaceLimit($domain),
                    'bandwidth_used'   => $domain->values->bandwidth_byte_usage[0] ?? 0,
                    'bandwidth_limit'   => $this->getBandwidthLimit($domain),
                    'status'           => (isset($domain->values->disabled) ? 'suspended' : 'active'),
                ];
            }
        }

        return $domains;
    }

    /**
     * Format the raw output from the Virtualmin list-plans program into a more user-friendly format.
     */
    public function getPlans(): array
    {
        $plans = [];

        $output = json_decode($this->listPlans());

        foreach ($output->data as $plan) {
            $plans[] = [
                'server'            => $this->server['hostname'],
                'remote_id'         => $plan->name,
                'name'              => $plan->values->name[0],
                'disk_space_limit'  => $plan->values->server_block_quota[0],
                'bandwidth_limited' => $plan->values->maximum_bw[0],
            ];
        }

        return $plans;
    }

    /**
     * Run the Virtualmin list-domains command and return the output.
     */
    public function listDomains()
    {
        $program = 'list-domains';

        return $this->runProgram($program);
    }

    /**
     * Run the Virtualmin list-domains command and return the output.
     */
    public function listPlans()
    {
        $program = 'list-plans';

        return $this->runProgram($program);
    }

    /**
     * Adds a new Virtualmin virtual server, with the settings and features specified on the command line.
     */
    public function createDomain($info = [])
    {
        $program = 'create-domain';

        return $this->runProgram($program, $info);
    }

    /**
     * Deletes an existing Virtualmin virtual server and all sub-servers, mailboxes and alias domains.
     */
    public function deleteDomain($info = [])
    {
        $program = 'delete-domain';

        return $this->runProgram($program, $info);
    }

    private function getBandwidthLimit($domain)
    {
        if (isset($domain->values->bandwidth_byte_limit)) {
            if ($domain->values->bandwidth_byte_limit[0] == 'Unlimited') {
                return -1;
            }

            return $domain->values->bandwidth_byte_limit[0];
        }

        return null;
    }

    private function getDiskSpaceLimit($domain)
    {
        if (isset($domain->values->server_block_quota)) {
            if ($domain->values->server_block_quota[0] == 'Unlimited') {
                return -1;
            }

            return $domain->values->server_block_quota[0];
        }

        return null;
    }

    /*
    * Hosting can be either top level hosting or a subdomain of a top level server
    */
    private function getHostingType($domain): ?string
    {
        if (isset($domain->values->type)) { // TODO PHP 8.0 provides shortcut for this
            if ($domain->values->type[0] == 'Top-level server') {
                return 'Web Hosting';
            }
            if ($domain->values->type[0] == 'Sub-server') {
                return 'Subdomain';
            }
        }

        return null;
    }

    public function runProgram($program, $data = null)
    {
        $hostname = $this->server['hostname'];
        $username = $this->server['username'] ?? 'root';
        $password = $this->server['password'];
        $port = $this->server['port'] ?? '10000';

        ray('runProgram server', $this->server);

        if ($this->mode == 'read_cache') {
            return file_get_contents('storage/'.$program.'.json');
        }
        $url = "https://$hostname/virtual-server/remote.cgi?json=1&multiline&program=$program";

        if (is_array($data) && ! empty($data)) {
            $url .= '&'.http_build_query($data);
        }

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_PORT, $port);
        curl_setopt($ch, CURLOPT_HTTPHEADER, [
            'Authorization: Basic '.base64_encode($username.':'.$password),
        ]);

        curl_setopt($ch, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_TIMEOUT, 1800);
        curl_setopt($ch, CURLOPT_VERBOSE, empty($this->quiet));

        ray()->measure();
        $result = curl_exec($ch);
        ray()->measure();

        if ($this->mode == 'write_cache' && $result) {
            file_put_contents('storage/'.$program.'.json', $result);
        }

        return $result;
    }
}
